test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,365 @@
+-- if server is already raft leader and limbo owner.
+g.before_test('test_leader_promote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    make_leader(g.server_1)
+end)
+
+g.test_leader_promote = function(g)
+    local ok, _ = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, 'error while promoting leader')
+end
+
+-- Test that box_promote will return 0
+-- if server is already raft leader but not current limbo owner.
+g.before_test('test_raft_leader_promote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+end)
+
+g.test_raft_leader_promote = function(g)
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_BOX_RAFT_SYNCHRO_QUEUE_DELAY', true)
+        box.ctl.promote()
+    end)
+    g.server_1:wait_election_state('leader')
+
+    local ok = g.server_1:exec(function()
+        local ok, _ = pcall(box.ctl.promote)
+        box.error.injection.set('ERRINJ_BOX_RAFT_SYNCHRO_QUEUE_DELAY', false)
+        return ok
+    end)
+    luatest.assert(ok, 'error while promoting raft leader')
+end
+
+-- Test that box_demote will return 0
+-- if server is already follower.
+g.before_test('test_follower_demote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    make_follower(g.server_1)
+end)
+
+g.test_follower_demote = function(g)
+    local ok, _ = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, 'error while demoting follower')
+end
+
+-- Test that box_promote will return box.error.UNSUPPORTED
+-- when called while in promote already
+g.test_simultaneous_promote = function(g)
+    local ok, err = g.server_1:exec(function()
+        require('fiber').create(box.ctl.promote)
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting while in promote')
+end
+
+-- Test that box_demote will return box.error.UNSUPPORTED
+-- when called while in promote already
+g.test_simultaneous_demote = function(g)
+    local ok, err = g.server_1:exec(function()
+        require('fiber').create(box.ctl.promote)
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while demoting while in promote')
+end
+
+-- Test that box_promote will return box.error.UNSUPPORTED
+-- when trying to promote voter
+g.before_test('test_voter_promote', function(g)
+    g.server_1:box_config({election_mode = 'voter'})
+end)
+
+g.test_voter_promote = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting voter')
+end
+
+-- Test interfering promotion for election_mode = 'off'
+-- while in WAL delay
+g.test_wal_interfering_promote = function(g)
```

Comment:
Let's unify this test with `test_wal_interfering_demote`.
They only differ in a single `promote/demote` call and can be parametrized.